tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/Services/WorkTokens/CosmosAsyncLockProvider.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.Azure.Cosmos;
+
+namespace Azure.Sdk.Tools.PipelineWitness.Services.WorkTokens
+{
+    public class CosmosAsyncLockProvider : IAsyncLockProvider
+    {
+        private readonly Container container;
+
+        public CosmosAsyncLockProvider(CosmosClient cosmosClient, string databaseName, string containerName)
+        {
+            if (cosmosClient == null)
+            {
+                throw new ArgumentNullException(nameof(cosmosClient));
+            }
+
+            this.container = cosmosClient.GetContainer(databaseName, containerName);
+        }
+
+        public async Task<IAsyncLock> GetLockAsync(string id, TimeSpan duration, CancellationToken cancellationToken)
+        {
+            var partitionKey = new PartitionKey(id);
+
+            ItemResponse<CosmosLockDocument> response;
+            
+            try
+            {
+                response = await this.container.ReadItemAsync<CosmosLockDocument>(id, partitionKey, cancellationToken: cancellationToken);
+            }
+            catch (CosmosException ex) when (ex.StatusCode == HttpStatusCode.NotFound)
+            {
+                return await CreateLockAsync(id, duration, cancellationToken);
+            }
+
+            var existingLock = response.Resource;
+
+            if (existingLock.Expiration >= DateTime.UtcNow)
+            {
+                return null;
```

Comment:
Stupid question, where is the mechanism for an instance to pick up work when a lock that is expired?